pennylane/beta/interfaces/autograd.py

```diff
@@ -0,0 +1,164 @@
+    """Mixin class for applying an autograd interface to a :class:`~.QuantumTape`.
+
+    Autograd-compatible quantum tape classes can be created via subclassing:
+
+    .. code-block:: python
+
+        class MyAutogradQuantumTape(AutogradInterface, QuantumTape):
+
+    Alternatively, the autograd interface can be dynamically applied to existing
+    quantum tapes via the :meth:`~.apply` class method. This modifies the
+    tape **in place**.
+
+    Once created, the autograd interface can be used to perform quantum-classical
+    differentiable programming.
+
+    .. note::
+
+        If using a device that supports native autograd computation and backpropagation, such as
+        :class:`~.DefaultQubitAutograd`, the Autograd interface **does not need to be applied**. It
+        is only applied to tapes executed on non-Autograd compatible devices.
+
+    **Example**
+
+    Once an autograd quantum tape has been created, it can be differentiated using autograd:
+
+    .. code-block:: python
+
+        tape = AutogradInterface.apply(QuantumTape())
+
+        with tape:
+            qml.Rot(0, 0, 0, wires=0)
+            expval(qml.PauliX(0))
+
+        def cost_fn(x, y, z, device):
+            tape.set_parameters([x, y ** 2, y * np.sin(z)], trainable_only=False)
+            return tape.execute(device=device)
+
+    >>> x = np.array(0.1, requires_grad=False)
+    >>> y = np.array(0.2, requires_grad=True)
+    >>> z = np.array(0.3, requires_grad=True)
+    >>> dev = qml.device("default.qubit", wires=2)
+    >>> cost_fn(x, y, z, device=dev)
+    [0.03991951]
+    >>> jac_fn = qml.jacobian(cost_fn)
+    >>> jac_fn(x, y, z, device=dev)
+    [[ 0.39828408 -0.00045133]]
+    """
+
+    # pylint: disable=attribute-defined-outside-init
+    dtype = np.float64
+
+    @property
+    def interface(self):  # pylint: disable=missing-function-docstring
+        return "autograd"
+
+    def _update_trainable_params(self):
+        params = []
+
+        for p_idx in self._par_info:
+            op = self._par_info[p_idx]["op"]
+            op_idx = self._par_info[p_idx]["p_idx"]
+            params.append(op.data[op_idx])
+
+        trainable_params = set()
+
+        for idx, p in enumerate(params):
+            if getattr(p, "requires_grad", True):
+                trainable_params.add(idx)
+
+        self.trainable_params = trainable_params
+        return params
+
+    def get_parameters(self, trainable_only=True):  # pylint: disable=missing-function-docstring
+        params = self._update_trainable_params()
+
+        if trainable_only:
+            params = [p for idx, p in enumerate(params) if idx in self.trainable_params]
+
+        return autograd.builtins.list(params)
+
+    @autograd.extend.primitive
+    def _execute(self, params, device):
+        params = autograd.builtins.tuple(params)
+
+        res = self.execute_device(params, device=device)
+
+        if res.dtype == np.dtype("object"):
+            return np.hstack(res)

```

Comment:
A NumPy object array simply refers to a NumPy array that stores a non-NumPy datatype :) In this case, it would likely be a list:

![image](https://user-images.githubusercontent.com/2959003/93406976-325c0980-f8d0-11ea-9dae-110a46a53600.png)

Any 'array' that is ragged will automatically be cast to an object array, since ragged arrays are not natively supported by NumPy. So you will still be able to use most NumPy functions on it, but NumPy will simply be storing it in memory as a Python nested list, rather than a more memory efficient fortran memory view.